torch/utils/data/datapipes/utils/decoder.py

```diff
@@ -279,7 +279,10 @@ def decode1(self, key, data):
 
         # if data is a stream handle, we need to read all the content before decoding
         if isinstance(data, io.BufferedIOBase) or isinstance(data, io.RawIOBase):
+            bs = data
             data = data.read()
+            # Explicitly close binary stream
+            bs.close()
```

Comment:
Close binary stream after read. Reduce warning from python gc.

Reply:
But shouldn't this happen where ever the stream is opened? IMO we should yield the stream and close it when we resume the datapipe. Plus, we should wrap the yield in a `try / finally` block to make sure the file handle will be closed regardless of what happens downstream. Better yet, open the files through the context manager syntax to not have to deal with this at all.

Reply:
>  IMO we should yield the stream and close it when we resume the datapipe.

I am not sure about it. What do you mean by resume.

> Better yet, open the files through the context manager syntax to not have to deal with this at all.

I have tried to use context manager to close the file. It's fine for most cases. But, if user starts to use `batch` over the each context manager. The `file` will be closed. We need to have a mechanism to prevent batching for these file handlers downstream.
```py
class DP:
    def __iter__(self):
        for file in self.files:
            with open(file) as f:
                yield f

dp = DP(files)
for f in dp:
    data = f.read()  # Working

dp = DP(files).batch(2)
for f in dp:
    data1 = f[0].read()  # ValueError: I/O operation on closed file.
```
Do you have any suggestion on this?

Reply:
Not sure what the `.batch()` call does here. Is this implemented somewhere?

Reply:
It's just a different syntax for a stack of `DataPipe`. `dp = DP(files).batch(2)` is same as:
```py
dp = DP(files)
dp = Batch(dp, batch_size=2)  # a batch of file handlers with `batch_size` 2
```

Reply:
I am sorry that I should explain more detail about it. I am using `batch` to mimic the `prefetch`. Considering we would have this feature, it may cause file closed after deplete the generator.